src/org/omegat/gui/exttrans/MachineTranslateTextArea.java

```diff
@@ -147,18 +205,34 @@ private void startSearchThread(SourceTextEntry newEntry, boolean force) {
     @Override
     protected void setFoundResult(final SourceTextEntry se, final MachineTranslationInfo data) {
         UIThreadsUtil.mustBeSwingThread();
-
         if (data != null && data.result != null) {
-            displayed.add (data);
-            setText(getText() + data.result + "\n<" + data.translatorName + ">\n\n");
+            String indexString = String.format("%d", index);
+            // build content
+            StringBuilder sb = new StringBuilder();
+            sb.append("<div id=\"").append(indexString).append("\">");
+            sb.append(data.result);
+            sb.append("<div class=\"engine\">&lt;");
+            sb.append(data.translatorName);
+            sb.append("&gt;</div></div>");
+            try {
+                HTMLDocument doc = (HTMLDocument) getDocument();
+                HTMLEditorKit kit = (HTMLEditorKit) getEditorKit();
+                kit.insertHTML(doc, doc.getLength(), sb.toString(), 0, 0, null);
+                displayed.put(data, indexString);
+                index++;
```

Comment:
I don't like having `index` as separate state in the instance. If I understand correctly, `index` will always be the same as `displayed.size()`.

In fact you don't need `displayed` to be a Map at all if the value is just the stringified index. You could just use a List and in `flashSelected` use `List#indexOf`. The list will almost always be so small that a linear search is plenty fast enough, so there is no need for a Set or Map.

Reply:
Adjusted.